src/rai_core/rai/agents/spatiotemporal/spatiotemporal_agent.py

```diff
@@ -0,0 +1,340 @@
+    x: float
+    y: float
+    z: float
+    w: float
+
+
+class Pose(BaseModel):
+    position: Point
+    orientation: Quaternion
+
+
+class PoseStamped(BaseModel):
+    header: Header
+    pose: Pose
+
+
+class SpatioTemporalData(BaseModel):
+    id: str = Field(default_factory=lambda: str(uuid.uuid4()))
+    timestamp: Annotated[float, "timestamp"]
+    images: Dict[Annotated[str, "camera topic"], str] = Field(repr=False)
+    tf: Optional[PoseStamped]
+    temporal_context: Annotated[str, "compressed history of messages"]
+    image_text_descriptions: Annotated[str, "text descriptions of images"]
+
+
+class SpatioTemporalConfig(BaseModel):
+    db_url: str
+    db_name: str
+    collection_name: str
+    image_to_text_model: BaseChatModel
+    context_compression_model: BaseChatModel
+    vector_db: VectorStore
+    time_interval: float
+
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+
+
+class SpatioTemporalAgent(BaseAgent):
+    def __init__(
+        self,
+        config: SpatioTemporalConfig,
+        image_description_prompt: str = IMAGE_DESCRIPTION_PROMPT,
+        context_compression_prompt: str = CONTEXT_COMPRESSION_PROMPT,
+        *args: tuple[Any, ...],
+        **kwargs: dict[str, Any],
+    ):
+        """
+        Initialize the SpatioTemporalAgent.
+
+        Parameters
+        ----------
+        config : SpatioTemporalConfig
+            Configuration for the spatiotemporal agent.
+        *args : tuple
+            Additional positional arguments.
+        **kwargs : dict
+            Additional keyword arguments.
+        """
+        super().__init__(*args, **kwargs)  # type: ignore
+        self.config = config
+        self.image_description_prompt = image_description_prompt
+        self.context_compression_prompt = context_compression_prompt
+        self.db = MongoClient(self.config.db_url)[self.config.db_name]  # type: ignore
+        self.collection = self.db[self.config.collection_name]  # type: ignore
+        self.logger = logging.getLogger(__name__)
+        self.thread: Optional[threading.Thread] = None
+        self._stop_event = threading.Event()
+
+    @abstractmethod
+    def _get_images(
+        self,
+    ) -> Dict[Annotated[str, "image source"], Annotated[str, "base64 encoded image"]]:
+        """
+        Abstract method to get images from sources (e.g. robot's cameras).
+
+        Returns
+        -------
+        Dict[Annotated[str, "image source"], Annotated[str, "base64 encoded image"]]
+            A dictionary mapping image sources to image data.
+        """
+        pass
+
+    @abstractmethod
+    def _get_tf(self) -> Optional[PoseStamped]:
+        """
+        Abstract method to get the transformation data (e.g. robot's pose).
+
+        Returns
+        -------
+        Optional[PoseStamped]
+            The transformation data, if available.
+        """
+        pass
+
+    @abstractmethod
+    def _get_robots_history(self) -> List[BaseMessage]:
```

Comment:
It suggests multiple `robots`
```suggestion
    def _get_robot_history(self) -> List[BaseMessage]:
```

Reply:
Wait, you are suggesting robots but the suggestion shows otherwise

Reply:
I know ;) But current name suggests more robots not robot's. I would make it get_robot_history for single, get_robots_history for multi robots. 
But it's also fine to keep you version if you prefer it